tests/integrationv2/test_external_psk.py

```diff
@@ -0,0 +1,294 @@
+                                       b"0727c621a79141ef", b"5f7de6505e5bfbc3", b"88e9334369409393",
+                                       b"4ae4d357" ]
+
+
+class Outcome(enum.Enum):
+    psk_connection = auto()
+    full_handshake = auto()
+    connection_failed = auto()
+
+S2N_TEST_CASE = namedtuple(
+    'S2N_TEST_CASE', 'cipher certificate s2nc_params s2nd_params outcome')
+
+S2N_TEST_SUITE = [
+    S2N_TEST_CASE(cipher=Ciphers.CHACHA20_POLY1305_SHA256,
+                  certificate=None,
+                  s2nd_params=[ '--psk', s2n_known_value_psk_sha256 ],
+                  s2nc_params=[ '--psk', s2n_known_value_psk_sha256 ],
+                  outcome=Outcome.psk_connection),
+    S2N_TEST_CASE(cipher=Ciphers.AES256_GCM_SHA384,
+                  certificate=None,
+                  s2nd_params=[ '--psk', s2n_known_value_psk_sha384 ],
+                  s2nc_params=[ '--psk', s2n_known_value_psk_sha384 ],
+                  outcome=Outcome.psk_connection),
+    S2N_TEST_CASE(cipher=Ciphers.AES128_GCM_SHA256,
+                  certificate=None,
+                  s2nd_params=[ '--psk', s2n_psk_parameters_sha384, 
+                                '--psk', s2n_known_value_psk_sha256 ],
+                  s2nc_params=[ '--psk', s2n_psk_parameters_sha256, 
+                                '--psk', s2n_known_value_psk_sha256 ],
+                  outcome=Outcome.psk_connection),
+    # Note that the following test case's outcome is `full_handshake` instead of `connection_failed` 
+    # when `certificate` is set to `None` because s2nd uses a default certificate whencle a certificate is not provided.
+    S2N_TEST_CASE(cipher=Ciphers.CHACHA20_POLY1305_SHA256,
+                  certificate=None,
+                  s2nd_params=[ '--psk', s2n_psk_parameters_sha384 ],
+                  s2nc_params=[ '--psk', s2n_psk_parameters_sha256 ],
+                  outcome=Outcome.full_handshake),
+    S2N_TEST_CASE(cipher=Ciphers.AES128_GCM_SHA256,
+                  certificate=Certificates.ECDSA_256,
+                  s2nd_params=[ '--psk', s2n_psk_parameters_sha384 ],
+                  s2nc_params=[ '--psk', s2n_psk_parameters_sha256 ],
+                  outcome=Outcome.full_handshake)
+]
+
+OPENSSL_TEST_CASE = namedtuple(
+    'OPENSSL_TEST_CASE', 'cipher certificate s2n_params openssl_params outcome')
+
+OPENSSL_SERVER_S2NC_TEST_SUITE = [
+    OPENSSL_TEST_CASE(cipher=Ciphers.AES128_GCM_SHA256,
+                      certificate=None,
+                      s2n_params=[ '--psk', s2n_known_value_psk_sha256 ],
+                      openssl_params=openssl_known_value_psk_parameters,
+                      outcome=Outcome.psk_connection),
+    OPENSSL_TEST_CASE(cipher=Ciphers.CHACHA20_POLY1305_SHA256,
+                      certificate=None,
+                      s2n_params=[ '--psk', s2n_known_value_psk_sha256 ],
+                      openssl_params=openssl_known_value_psk_parameters,
+                      outcome=Outcome.psk_connection),
```

Comment:
Rather than manually parameterizing the cipher, could you just use the standard fixtures and ignore any combinations with the wrong hash?

Reply:
The manual parameterizing was done as psk tests are special cased for openssl server and client - they don't behave in the same manner s2nd/c does. Making them general will include multiple if-else's and making it complex to follow and maintain.